# Bayonne Medical Center
https://carepointhealth.org/patients-visitors/hospital-charges/
https://carepointhealth.org/wp-content/uploads/2023/01/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv

In [2]:
# Load duckdb, which lets us efficiently load large files
import duckdb

# Load pandas, which lets us manipulate dataframes
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

# Set configrations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.
%config SqlMagic.autopandas = True

%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

# Allow named parameters (python variables) in SQL cells
%config SqlMagic.named_parameters=True

# Connect jupysql to DuckDB using a SQLAlchemy-style connection string. Either connect to an in memory DuckDB, or a file backed db.
%sql duckdb:///:memory:

In [5]:
!wget  --header="Accept: text/html" --user-agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:21.0) Gecko/20100101 Firefox/21.0" https://carepointhealth.org/wp-content/uploads/2023/01/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv -P /tmp

--2023-09-22 13:43:49--  https://carepointhealth.org/wp-content/uploads/2023/01/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv
Resolving carepointhealth.org (carepointhealth.org)... 46.21.144.106
Connecting to carepointhealth.org (carepointhealth.org)|46.21.144.106|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1334141 (1.3M) [text/csv]
Saving to: ‘/tmp/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv’

261442063_BAYONNEME 100%[===================>]   1.27M  1.62MB/s    in 0.8s    

2023-09-22 13:43:50 (1.62 MB/s) - ‘/tmp/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv’ saved [1334141/1334141]



In [6]:
!head /tmp/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv

Bayonne Medical Center,,,,,,,,,,,,,,,,,,,,,,,,
Standard Charges,,,,,,,,,,,,,,,,,,,,,,,,
Filename: 261442063_BAYONNEMEDICALCENTER_STANDARD CHARGES,,,,,,,,,,,,,,,,,,,,,,,,
Updated 12/30/2021,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,
Charge Code,Charge Description,Department Name,Place of Service, Charge , Cash Price , Horizon PPO , Horizon HMO , Horizon Indemnity , Horizon Medicare , United Healthcare , Oxford , United Medicare , Clover Medicare , Wellcare Medicare , Wellcare Medicaid , Horizon NJ Health , AmeriGroup Medicare , Amerigroup Medicaid , Cigna , Aetna , Aetna Medicare , Aetna Medicaid , Min Negotiated Charge , Max Negotiated Charge 
10008,LEAVE OF ABSENCE,PSYCHIATRIC ,Inpatient, $0.01 , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 ", N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 "," $1,250.00 "
100015,ROOM & BOARD M/S,ICU/CCU,Inpatient," $18,000.00 ", N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 ", N/A , 

# Claude Prompt

````
```
%%sql
SELECT *
FROM read_csv('https://data.cityofnewyork.us/api/views/erm2-nwe9/rows.csv?accessType=DOWNLOAD',
    header=True,
    delim=',',
    quote='"',
    columns={'Unique Key': 'BIGINT',
    'Created Date': 'VARCHAR',
    'Closed Date': 'VARCHAR',
    'Agency': 'VARCHAR',
    'Agency Name': 'VARCHAR',
    'Complaint Type': 'VARCHAR',
    'Descriptor': 'VARCHAR',
    'Location Type': 'VARCHAR',
    'Incident Zip': 'VARCHAR',
    'Incident Address': 'VARCHAR',
    'Street Name': 'VARCHAR',
    'Cross Street 1': 'VARCHAR',
    'Cross Street 2': 'VARCHAR',
    'Intersection Street 1': 'VARCHAR',
    'Intersection Street 2': 'VARCHAR',
    'Address Type': 'VARCHAR',
    'City': 'VARCHAR',
    'Landmark': 'VARCHAR',
    'Facility Type': 'VARCHAR',
    'Status': 'VARCHAR',
    'Due Date': 'VARCHAR',
    'Resolution Description': 'VARCHAR',
    'Resolution Action Updated Date': 'VARCHAR',
    'Community Board': 'VARCHAR',
    'BBL': 'VARCHAR',
    'Borough': 'VARCHAR',
    'X Coordinate (State Plane)': 'VARCHAR',
    'Y Coordinate (State Plane)': 'VARCHAR',
    'Open Data Channel Type': 'VARCHAR',
    'Park Facility Name': 'VARCHAR',
    'Park Borough': 'VARCHAR',
    'Vehicle Type': 'VARCHAR',
    'Taxi Company Borough': 'VARCHAR',
    'Taxi Pick Up Location': 'VARCHAR',
    'Bridge Highway Name': 'VARCHAR',
    'Bridge Highway Direction': 'VARCHAR',
    'Road Ramp': 'VARCHAR',
    'Bridge Highway Segment': 'VARCHAR',
    'Latitude': 'DOUBLE',
    'Longitude': 'DOUBLE',
    'Location': 'VARCHAR'}) 
LIMIT 10;
```

Please take the above SQL query and rewrite for the file `131624096_mount-sinai-hospital_standardcharges.txt` with the following header:

```
Bayonne Medical Center,,,,,,,,,,,,,,,,,,,,,,,,
Standard Charges,,,,,,,,,,,,,,,,,,,,,,,,
Filename: 261442063_BAYONNEMEDICALCENTER_STANDARD CHARGES,,,,,,,,,,,,,,,,,,,,,,,,
Updated 12/30/2021,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,
Charge Code,Charge Description,Department Name,Place of Service, Charge , Cash Price , Horizon PPO , Horizon HMO , Horizon Indemnity , Horizon Medicare , United Healthcare , Oxford , United Medicare , Clover Medicare , Wellcare Medicare , Wellcare Medicaid , Horizon NJ Health , AmeriGroup Medicare , Amerigroup Medicaid , Cigna , Aetna , Aetna Medicare , Aetna Medicaid , Min Negotiated Charge , Max Negotiated Charge 
10008,LEAVE OF ABSENCE,PSYCHIATRIC ,Inpatient, $0.01 , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 ", N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 "," $1,250.00 "
100015,ROOM & BOARD M/S,ICU/CCU,Inpatient," $18,000.00 ", N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 ", N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 "," $1,250.00 "
100016,ROOM & BOARD 2A,MED/SURG R&B,Inpatient," $18,000.00 ", N/A , N/A , N/A , N/A , N/A ," $3,182.00 "," $5,322.26 ", N/A , N/A , N/A , N/A ," $1,250.00 ", N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 "," $5,322.26 "
100024,ROOM & BOARD 2R,MED/SURG R&B,Inpatient," $18,000.00 ", N/A , N/A , N/A , N/A , N/A ," $3,182.00 "," $5,322.26 ", N/A , N/A , N/A , N/A ," $1,250.00 ", N/A , N/A , N/A , N/A , N/A , N/A ," $1,250.00 "," $5,322.26 "
```
````

In [17]:
%%sql 

SELECT * 
FROM read_csv('/tmp/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv', 
  header=True, 
  delim=',',
  quote='"',
  skip=5,
  nullstr=' N/A ',
  columns={'Charge Code': 'VARCHAR',  
           'Charge Description': 'VARCHAR',
           'Department Name': 'VARCHAR',
           'Place of Service': 'VARCHAR',
           'Charge': 'VARCHAR',
           'Cash Price': 'VARCHAR',
           'Horizon PPO': 'VARCHAR',
           'Horizon HMO': 'VARCHAR',
           'Horizon Indemnity': 'VARCHAR',
           'Horizon Medicare': 'VARCHAR',
           'United Healthcare': 'VARCHAR',
           'Oxford': 'VARCHAR',
           'United Medicare': 'VARCHAR',
           'Clover Medicare': 'VARCHAR',
           'Wellcare Medicare': 'VARCHAR',
           'Wellcare Medicaid': 'VARCHAR',
           'Horizon NJ Health': 'VARCHAR',
           'AmeriGroup Medicare': 'VARCHAR',
           'Amerigroup Medicaid': 'VARCHAR', 
           'Cigna': 'VARCHAR',
           'Aetna': 'VARCHAR',
           'Aetna Medicare': 'VARCHAR',
           'Aetna Medicaid': 'VARCHAR',
           'Min Negotiated Charge': 'VARCHAR',
           'Max Negotiated Charge': 'VARCHAR'})
LIMIT 10;

Charge Code      Charge Description    Department Name Place of Service  \
0       10008        LEAVE OF ABSENCE       PSYCHIATRIC         Inpatient   
1      100015        ROOM & BOARD M/S            ICU/CCU        Inpatient   
2      100016         ROOM & BOARD 2A       MED/SURG R&B        Inpatient   
3      100024         ROOM & BOARD 2R       MED/SURG R&B        Inpatient   
4      100057         ROOM & BOARD 3R       MED/SURG R&B        Inpatient   
5      100073         ROOM & BOARD 5R       MED/SURG R&B        Inpatient   
6      100075       ROOM & BOARD TRCU  TRANSITIONAL CARE        Inpatient   
7      100081      ROOM & BOARD PSYCH       PSYCHIATRIC         Inpatient   
8      100091  ROOM & BOARD TELEMETRY       MED/SURG R&B        Inpatient   
9      100095   ROOM & BOARD MED/SURG       MED/SURG R&B        Inpatient   

         Charge Cash Price Horizon PPO Horizon HMO Horizon Indemnity  \
0        $0.01         NaN         NaN         NaN               NaN   
1   $18,000.00         NaN         NaN         NaN               NaN   
2   $18,000.00         NaN         NaN         NaN               NaN   
3   $18,000.00         NaN         NaN         NaN               NaN   
4   $18,000.00         NaN         NaN         NaN               NaN   
5   $18,000.00         NaN         NaN         NaN               NaN   
6   $18,000.00         NaN         NaN         NaN               NaN   
7   $18,000.00         NaN         NaN         NaN               NaN   
8   $18,000.00         NaN         NaN         NaN               NaN   
9   $18,000.00         NaN         NaN         NaN               NaN   

  Horizon Medicare  ... Wellcare Medicaid Horizon NJ Health  \
0              NaN  ...               NaN        $1,250.00    
1              NaN  ...               NaN        $1,250.00    
2              NaN  ...               NaN        $1,250.00    
3              NaN  ...               NaN        $1,250.00    
4              NaN  ...               NaN        $1,250.00    
5              NaN  ...               NaN        $1,250.00    
6              NaN  ...               NaN        $1,250.00    
7              NaN  ...               NaN        $1,250.00    
8              NaN  ...               NaN        $1,250.00    
9              NaN  ...               NaN        $1,250.00    

  AmeriGroup Medicare Amerigroup Medicaid Cigna Aetna Aetna Medicare  \
0                 NaN                 NaN   NaN   NaN            NaN   
1                 NaN                 NaN   NaN   NaN            NaN   
2                 NaN                 NaN   NaN   NaN            NaN   
3                 NaN                 NaN   NaN   NaN            NaN   
4                 NaN                 NaN   NaN   NaN            NaN   
5                 NaN                 NaN   NaN   NaN            NaN   
6                 NaN                 NaN   NaN   NaN            NaN   
7                 NaN                 NaN   NaN   NaN            NaN   
8                 NaN                 NaN   NaN   NaN            NaN   
9                 NaN                 NaN   NaN   NaN            NaN   

  Aetna Medicaid Min Negotiated Charge Max Negotiated Charge  
0            NaN            $1,250.00             $1,250.00   
1            NaN            $1,250.00             $1,250.00   
2            NaN            $1,250.00             $5,322.26   
3            NaN            $1,250.00             $5,322.26   
4            NaN            $1,250.00             $5,322.26   
5            NaN            $1,250.00             $5,322.26   
6            NaN            $1,250.00             $5,322.26   
7            NaN            $1,250.00             $5,322.26   
8            NaN            $1,250.00             $5,322.26   
9            NaN            $1,250.00             $5,322.26   

[10 rows x 25 columns]

In [18]:
%%sql 
COPY (
SELECT * 
FROM read_csv('/tmp/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.csv', 
  header=True, 
  delim=',',
  quote='"',
  skip=5,
  nullstr=' N/A ',
  columns={'Charge Code': 'VARCHAR',  
           'Charge Description': 'VARCHAR',
           'Department Name': 'VARCHAR',
           'Place of Service': 'VARCHAR',
           'Charge': 'VARCHAR',
           'Cash Price': 'VARCHAR',
           'Horizon PPO': 'VARCHAR',
           'Horizon HMO': 'VARCHAR',
           'Horizon Indemnity': 'VARCHAR',
           'Horizon Medicare': 'VARCHAR',
           'United Healthcare': 'VARCHAR',
           'Oxford': 'VARCHAR',
           'United Medicare': 'VARCHAR',
           'Clover Medicare': 'VARCHAR',
           'Wellcare Medicare': 'VARCHAR',
           'Wellcare Medicaid': 'VARCHAR',
           'Horizon NJ Health': 'VARCHAR',
           'AmeriGroup Medicare': 'VARCHAR',
           'Amerigroup Medicaid': 'VARCHAR', 
           'Cigna': 'VARCHAR',
           'Aetna': 'VARCHAR',
           'Aetna Medicare': 'VARCHAR',
           'Aetna Medicaid': 'VARCHAR',
           'Min Negotiated Charge': 'VARCHAR',
           'Max Negotiated Charge': 'VARCHAR'})
) TO '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.parquet' (COMPRESSION ZSTD);

Empty DataFrame
Columns: [Success]
Index: []

In [20]:
%%sql 
WITH unpivot_alias AS (
    UNPIVOT '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.parquet'
    ON COLUMNS(* EXCLUDE ('Charge Code', 'Charge Description', 'Department Name', 'Place of Service'))
    INTO
        NAME payor
        VALUE negotiated_rate
) 
SELECT * FROM unpivot_alias;

Charge Code              Charge Description Department Name  \
0           10008                LEAVE OF ABSENCE    PSYCHIATRIC    
1           10008                LEAVE OF ABSENCE    PSYCHIATRIC    
2           10008                LEAVE OF ABSENCE    PSYCHIATRIC    
3           10008                LEAVE OF ABSENCE    PSYCHIATRIC    
4          100015                ROOM & BOARD M/S         ICU/CCU   
...           ...                             ...             ...   
52646   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
52647   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
52648   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
52649   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
52650   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   

           Place of Service                  payor negotiated_rate  
0                 Inpatient                 Charge          $0.01   
1                 Inpatient      Horizon NJ Health      $1,250.00   
2                 Inpatient  Min Negotiated Charge      $1,250.00   
3                 Inpatient  Max Negotiated Charge      $1,250.00   
4                 Inpatient                 Charge     $18,000.00   
...                     ...                    ...             ...  
52646  Inpatient/Outpatient                  Aetna      $2,059.15   
52647  Inpatient/Outpatient         Aetna Medicare        $635.54   
52648  Inpatient/Outpatient         Aetna Medicaid         $78.88   
52649  Inpatient/Outpatient  Min Negotiated Charge         $78.88   
52650  Inpatient/Outpatient  Max Negotiated Charge      $2,329.89   

[52651 rows x 6 columns]

In [5]:
%%sql 
COPY (
WITH unpivot_alias AS (
    UNPIVOT '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES.parquet'
    ON COLUMNS(* EXCLUDE ('Charge Code', 'Charge Description', 'Department Name', 'Place of Service'))
    INTO
        NAME payor
        VALUE negotiated_rate
) 
SELECT * FROM unpivot_alias
WHERE payor != 'Charge'
AND payor != 'Place of Service'
AND payor != 'Department Name'
AND payor != 'Max Negotiated Charge'
AND payor != 'Min Negotiated Charge'
) TO '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES-unpivoted.parquet' (COMPRESSION ZSTD);

Empty DataFrame
Columns: [Success]
Index: []

In [6]:
%%sql 
SELECT * FROM '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES-unpivoted.parquet'; 

Charge Code              Charge Description Department Name  \
0           10008                LEAVE OF ABSENCE    PSYCHIATRIC    
1          100015                ROOM & BOARD M/S         ICU/CCU   
2          100016                 ROOM & BOARD 2A    MED/SURG R&B   
3          100016                 ROOM & BOARD 2A    MED/SURG R&B   
4          100016                 ROOM & BOARD 2A    MED/SURG R&B   
...           ...                             ...             ...   
34487   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
34488   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
34489   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
34490   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   
34491   94064017R  NEEDLE LOC US GUIDE 1ST LESION     MAMMOGRAPHY   

           Place of Service                payor negotiated_rate  
0                 Inpatient    Horizon NJ Health      $1,250.00   
1                 Inpatient    Horizon NJ Health      $1,250.00   
2                 Inpatient    United Healthcare      $3,182.00   
3                 Inpatient               Oxford      $5,322.26   
4                 Inpatient    Horizon NJ Health      $1,250.00   
...                     ...                  ...             ...  
34487  Inpatient/Outpatient  Amerigroup Medicaid         $78.88   
34488  Inpatient/Outpatient                Cigna      $2,224.39   
34489  Inpatient/Outpatient                Aetna      $2,059.15   
34490  Inpatient/Outpatient       Aetna Medicare        $635.54   
34491  Inpatient/Outpatient       Aetna Medicaid         $78.88   

[34492 rows x 6 columns]

In [3]:
%%sql 
SELECT "Charge Code" as billing_code,
        "Charge Description" as description,
        payor,
        CAST(regexp_replace(NULLIF(negotiated_rate, ''), '[$,]', '', 'g') AS DOUBLE)
FROM '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES-unpivoted.parquet'

billing_code                     description                payor  \
0            10008                LEAVE OF ABSENCE    Horizon NJ Health   
1           100015                ROOM & BOARD M/S    Horizon NJ Health   
2           100016                 ROOM & BOARD 2A    United Healthcare   
3           100016                 ROOM & BOARD 2A               Oxford   
4           100016                 ROOM & BOARD 2A    Horizon NJ Health   
...            ...                             ...                  ...   
34487    94064017R  NEEDLE LOC US GUIDE 1ST LESION  Amerigroup Medicaid   
34488    94064017R  NEEDLE LOC US GUIDE 1ST LESION                Cigna   
34489    94064017R  NEEDLE LOC US GUIDE 1ST LESION                Aetna   
34490    94064017R  NEEDLE LOC US GUIDE 1ST LESION       Aetna Medicare   
34491    94064017R  NEEDLE LOC US GUIDE 1ST LESION       Aetna Medicaid   

       CAST(regexp_replace(nullif(negotiated_rate, ''), '[$,]', '', 'g') AS DOUBLE)  
0                                                1250.00                             
1                                                1250.00                             
2                                                3182.00                             
3                                                5322.26                             
4                                                1250.00                             
...                                                  ...                             
34487                                              78.88                             
34488                                            2224.39                             
34489                                            2059.15                             
34490                                             635.54                             
34491                                              78.88                             

[34492 rows x 4 columns]

In [4]:
%%sql 
COPY (
SELECT "Charge Code" as billing_code,
        "Charge Description" as description,
        payor,
        CAST(regexp_replace(NULLIF(negotiated_rate, ''), '[$,]', '', 'g') AS DOUBLE) AS negotiated_rate
FROM '/Users/me/data/payless_health/261442063_BAYONNEMEDICALCENTER_STANDARD_CHARGES-unpivoted.parquet'
) TO '/Users/me/data/payless_health/261442063_bayonne-medical-center.parquet' (COMPRESSION ZSTD);

Empty DataFrame
Columns: [Success]
Index: []

In [5]:
!cp /Users/me/data/payless_health/261442063_bayonne-medical-center.parquet /Users/me/projects/payless.health/docs/public/data/